lib/completion.dart

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library dartpad.completion;
+
+import 'dart:async';
+import 'dart:convert' show JSON;
+
+import 'package:logging/logging.dart';
+
+import 'dartservices_client/v1.dart';
+import 'editing/editor.dart';
+import 'src/util.dart';
+
+Logger _logger = new Logger('completion');
+
+// TODO: For CodeMirror, we get a request each time the user hits a key when the
+// completion popup is open. We need to cache the results when appropriate.
+
+class DartCompleter extends CodeCompleter {
+  final DartservicesApi servicesApi;
+  final Document document;
+
+  CancellableCompleter _lastCompleter;
+
+  DartCompleter(this.servicesApi, this.document);
+
+  Future<List<Completion>> complete(Editor editor) {
+    // Cancel any open completion request.
+    if (_lastCompleter != null) _lastCompleter.cancel();
+
+    int offset = editor.document.indexFromPos(editor.document.cursor);
+
+    var request = new SourceRequest()
+      ..source = editor.document.value
+      ..offset = offset;
+
+    Stopwatch timer = new Stopwatch()..start();
+
+    CancellableCompleter completer = new CancellableCompleter();
+    _lastCompleter = completer;
+
+    servicesApi.complete(request).then((CompleteResponse response) {
+      if (completer.isCancelled) return;
+
+      _logger.info('completion request in ${timer.elapsedMilliseconds}ms; '
+          '${response.completions.length} completions, '
+          'offset=${response.replacementOffset}, '
+          'length=${response.replacementLength}');
+
+      List<AnalysisCompletion> analysisCompletions = response.completions.map(
+          (completion) {
+        return new AnalysisCompletion(
+            response.replacementOffset, response.replacementLength, completion);
+      }).toList();
+
+      int delta = offset - response.replacementOffset;
+
+      List<Completion> completions =  analysisCompletions.map((completion) {
+        // TODO: Move to using a LabelProvider; decouple the data from the
+        // rendering.
+        String displayString = completion.isMethod ? '${completion.text}()' : completion.text;
+        if (completion.returnType != null) displayString += ': ${completion.returnType}';
```

Comment:
We've been trying use '→' consistently for the return type instead of ':', that's what the new DartDoc will also use so it would be great if we could do the same here